src/engine/guidance/collapse_turns.cpp

```diff
@@ -0,0 +1,271 @@
+#include "engine/guidance/collapse_turns.hpp"
+#include "extractor/guidance/turn_instruction.hpp"
+#include "engine/guidance/collapse_scenario_detection.hpp"
+#include "engine/guidance/collapsing_utility.hpp"
+#include "util/bearing.hpp"
+#include "util/guidance/name_announcements.hpp"
+
+#include "util/debug.hpp"
+
+#include <cstddef>
+
+#include <boost/assert.hpp>
+#include <boost/optional.hpp>
+#include <iterator>
+
+using osrm::extractor::guidance::TurnInstruction;
+using osrm::util::angularDeviation;
+using namespace osrm::extractor::guidance;
+
+namespace osrm
+{
+namespace engine
+{
+namespace guidance
+{
+
+using RouteSteps = std::vector<RouteStep>;
+using RouteStepIterator = typename RouteSteps::iterator;
+namespace
+{
+const constexpr double MAX_COLLAPSE_DISTANCE = 30;
+
+// TODO rework this
+double findTotalTurnAngle(const RouteStep &entry_step, const RouteStep &exit_step)
+{
+    const auto exit_intersection = exit_step.intersections.front();
+    const auto exit_step_exit_bearing = exit_intersection.bearings[exit_intersection.out];
+    const auto exit_step_entry_bearing =
+        util::reverseBearing(exit_intersection.bearings[exit_intersection.in]);
+
+    const auto entry_intersection = entry_step.intersections.front();
+    const auto entry_step_entry_bearing =
+        util::reverseBearing(entry_intersection.bearings[entry_intersection.in]);
+    const auto entry_step_exit_bearing = entry_intersection.bearings[entry_intersection.out];
+
+    const auto exit_angle =
+        util::angleBetweenBearings(exit_step_entry_bearing, exit_step_exit_bearing);
+    const auto entry_angle =
+        util::angleBetweenBearings(entry_step_entry_bearing, entry_step_exit_bearing);
+
+    const double total_angle =
+        util::angleBetweenBearings(entry_step_entry_bearing, exit_step_exit_bearing);
+    // We allow for minor deviations from a straight line
+    if (((entry_step.distance < MAX_COLLAPSE_DISTANCE && exit_step.intersections.size() == 1) ||
+         (entry_angle <= 185 && exit_angle <= 185) || (entry_angle >= 175 && exit_angle >= 175)) &&
+        angularDeviation(total_angle, 180) > 20)
+    {
+        // both angles are in the same direction, the total turn gets increased
+        //
+        // a ---- b
+        //           \
```

Comment:
Not really a code change, but for `collapse_turns.cpp:61:9: warning: multi-line comment [-Wcomment]` it is possible to put a NO-BREAK SPACE character 0xA0 ' '. It will not be removed as a trailing white space, is invisible and suppresses the warning. The same for `src/engine/guidance/post_processing.cpp:343:9: warning: multi-line comment [-Wcomment]`